### PR TITLE
`HeteroData.subgraph()`

### LINKED ISSUE #4001: `HeteroData.subgraph()`
### 🚀 The feature, motivation and pitch

Similar to `Data.subgraph()`, there should exist a `HeteroData.subgraph()` method to compute subgraphs in a heterogeneous graph setting, *e.g.*, for obtaining inductive node splits. Here, `mask`/`index` should be of type `dict`, holding masks/indices for each/a subset of node types:
```python
hetero_data.subgraph({'paper': mask})
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wsad1**
Closes #4001.

This PR adds `subgraph` to `HeteroData`.

### CODE REVIEW

File: torch_geometric/data/hetero_data.py [hunk 1]
```diff
@@ -437,6 +437,79 @@ def rename(self, name: NodeType, new_name: NodeType) -> 'HeteroData':
...
+            >>> HeteroData(
+                paper={ x=[10, 16] },
+                author={ x=[5, 32] },
+                conference={ x=[5, 8] },
+                (paper, cites, paper)={ edge_index=[2, 50] },
+                (author, to, paper)={ edge_index=[2, 30] },
+                (paper, to, conference)={ edge_index=[2, 25] }
+                )
+
+            subset ={'paper': torch.tensor([3,4,5,6]),
```
Comment (Padarn) [inline torch_geometric/data/hetero_data.py:463]:
```suggestion
            subset = {'paper': torch.tensor([3,4,5,6]),
```

File: torch_geometric/data/hetero_data.py [hunk 2]
```diff
@@ -437,6 +437,79 @@ def rename(self, name: NodeType, new_name: NodeType) -> 'HeteroData':
...
+            print(data.subgraph(subset))
+            >>> HeteroData(
+                paper={ x=[4, 16] },
+                author={ x=[2, 32] },
+                (paper, cites, paper)={ edge_index=[2, 24] },
+                (author, to, paper)={ edge_index=[2, 5] }
+                )
+
+        Args:
+            subset (Dict[str, Tensor or BoolTensor]): A dictonary holding the
```
Comment (Padarn) [inline torch_geometric/data/hetero_data.py:475]: Do you think its worth allowing also `List[str]` when you just want to filter node types out?

File: torch_geometric/data/hetero_data.py [hunk 3]
```diff
@@ -437,6 +437,79 @@ def rename(self, name: NodeType, new_name: NodeType) -> 'HeteroData':
...
+        """
+
+        data = HeteroData(self._global_store)
+
+        for node_type in self.node_types:
+            if node_type not in subset:
+                continue
+            cur_num_nodes = self[node_type].num_nodes
+            for key, value in self[node_type].items():
+                if isinstance(value,
```
Comment (Padarn) [inline torch_geometric/data/hetero_data.py:486]: maybe you could use the `is_node_attr` and `is_edge_attr` to simplify this loop quite a bit?

File: torch_geometric/data/hetero_data.py (hunk 2 above)
Comment (rusty1s) [reply torch_geometric/data/hetero_data.py:475]: Good idea. Let's do this in a separate PR since I am not sure we want to re-use `subgraph` for this due its expensive re-labeling computation.

**Review by rusty1s — APPROVED** [review]
Thank you (and welcome back :))
